x-pack/plugins/fleet/public/applications/fleet/sections/agent_policy/components/agent_policy_advanced_fields/custom_fields/global_data_tags_table.tsx

```diff
@@ -0,0 +1,394 @@
+  AgentPolicy,
+  GlobalDataTag,
+} from '../../../../../../../../common/types';
+
+interface Props {
+  updateAgentPolicy: (u: Partial<NewAgentPolicy | AgentPolicy>) => void;
+  initialTags: GlobalDataTag[];
+}
+
+export const GlobalDataTagsTable: React.FunctionComponent<Props> = ({
+  updateAgentPolicy,
+  initialTags,
+}) => {
+  const [globalDataTags, setGlobalDataTags] = useState<GlobalDataTag[]>(initialTags);
+  const [editIndexList, setEditIndexList] = useState<Set<number>>(new Set([]));
+  // isAdding indicates whether a new row is currently being added to the table.
+  // When true, the table will display "Confirm" and "Cancel" actions for the new row
+  // to allow the user to either confirm the addition or cancel it.
+  const [isAdding, setIsAdding] = useState<boolean>(false);
+  const [newTag, setNewTag] = useState<GlobalDataTag>({ name: '', value: '' });
+  const [newTagErrors, setNewTagErrors] = useState<{ name: string; value: string }>({
+    name: '',
+    value: '',
+  });
+  const [errors, setErrors] = useState<
+    Record<number, { name: string | null; value: string | null }>
+  >({});
+  const [isModalVisible, setIsModalVisible] = useState(false);
+  const [deleteIndex, setDeleteIndex] = useState<number | null>(null);
+
+  // Necessary to remove uncommitted tags after save is cancelled
+  useEffect(() => {
+    setGlobalDataTags(initialTags);
+  }, [initialTags]);
+
+  const handleAddField = () => {
+    setIsAdding(true);
+    setNewTag({ name: '', value: '' });
+  };
+
+  const handleEditChange = (e: React.ChangeEvent<HTMLInputElement>, index: number) => {
+    const updatedTags = globalDataTags.map((tag, i) =>
+      i === index ? { ...tag, [e.target.name]: e.target.value } : tag
+    );
+    setGlobalDataTags(updatedTags);
+    setErrors((prevErrors) => ({
+      ...prevErrors,
+      [index]: { ...prevErrors[index], [e.target.name]: null },
+    }));
+  };
+
+  const handleNewTagChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    setNewTag({
+      ...newTag,
+      [e.target.name]: e.target.value,
+    });
+    setNewTagErrors({ ...newTagErrors, [e.target.name]: null });
+  };
+
+  const validateTag = (tag: GlobalDataTag, index?: number) => {
+    const trimmedName = tag.name.trim();
+    const trimmedValue = tag.value.toString().trim();
+
+    let nameError = '';
+    let valueError = '';
+
+    if (!trimmedName) {
+      nameError = 'Name cannot be empty';
+    } else if (/\s/.test(trimmedName)) {
+      nameError = 'Name cannot contain spaces';
+    } else if (globalDataTags.some((t, i) => i !== index && t.name === trimmedName)) {
+      nameError = 'Name must be unique';
+    }
+
+    if (!trimmedValue) {
+      valueError = 'Value cannot be empty';
+    }
+
+    return { nameError, valueError, isValid: !nameError && !valueError };
+  };
+
+  const parseValue = (value: string): string | number => {
+    const parsedValue = Number(value);
+    return isNaN(parsedValue) ? value : parsedValue;
```

Comment:
Could you clarify what you want to check here? AFAIK casting an empty string to number for instance results in 0, so then `isNaN(0) = true`. Is that what you are tring to achieve?

In general the [isNaN function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/isNaN) has a bit of "unexpected" results, so rather than relying on it to check if the value is a number or a string, I would prefer to check the type beforehand. Something like
```
if (typeof value === "string") {
}

// or 
if (typeof value === "number") {
}
```

Reply:
Here I'm trying to preserve the input types by checking if the input can be parsed as a number. The type for all the text field inputs is a string so checking the type won't help here. What I want to do is if a user provides a string such as "123" I want this input to be sent as a number in the payload and not as a string. The  input for this function will not be an empty string at any time because the validation step rejects empty strings before getting to this function. 
When it comes to isNaN, isNaN(0) gives false, but as I mentioned, the function will never have an empty string as input.

Reply:
What I mean is that you can re-write the `parseValue` function to accept both types:
``` ts
  const parseValue = (value: string | number): string | number => {
    if (typeof value === 'string') {
      return value.trim();
    } else if (typeof value === 'number') {
      return value;
    }
    return '';
  };
```

and in all the places where you call it you can pass directly `value`, without need to cast it to a string:
``` ts
value: parseValue(newTag.value)
```